state.go

```diff
@@ -207,19 +214,23 @@ func DefaultUser(u string) llb.StateOption {
 func AddDefaultEnv(key, value string) llb.StateOption {
 	return func(st llb.State) llb.State {
 		st = st.AddEnv(key, value)
-		return withImageConfigMutator(st, func(_ context.Context, img *mdispec.DockerOCIImage) error {
+		return withImageConfigMutator(st, func(ctx context.Context, img *ImageConfig) error {
 			img.Config.Env = append(img.Config.Env, key+"="+value)
 			// In Dockerfile, multiple ENV can be specified in the same ENV command
 			// leading to one history element. This checks if the previous history
 			// committed was also an ENV, in which case it should just add to the
 			// previous history element.
+			// The sentinelTime is to ensure we only append to history records
+			// created in this session.
+			sentinelTime := sentinelHistoryTime(ctx)
 			numHistory := len(img.History)
-			if numHistory > 0 && strings.HasPrefix(img.History[numHistory-1].CreatedBy, "ENV") {
+			if numHistory > 0 && strings.HasPrefix(img.History[numHistory-1].CreatedBy, "ENV") && img.History[numHistory-1].Created == sentinelTime {
```

Comment:
This looks like a pointer comparison. I think that's the intent (implied by "sentinel" time), but it might be a good idea to document this on `sentinelHistoryTime`. It feels a bit fragile that something could unintentionally mutate the sentinel time value, and also fragile that this wouldn't be robust to deep copies, but I suppose the alternative would be to have some kind of mechanism tracking the indices of history entries added during the current session, and I don't have a full picture of how much complexity that would add.

Reply:
I ended up modifying the config data so we could track history with a sessionID, then added a sessionID context value to let us decide on insert/mutate actions.